request_handlers.cc

```diff
@@ -30,14 +39,120 @@ void EchoRequestHandler::handleRequest(const char* request_buffer,
-  // TODO: Allocate enough memory for whole file and return pointer to file
-  // TODO: Fill out response_buffer with file data
+  // Boost has a great tokenizer
+  // See: https://stackoverflow.com/questions/53849/how-do-i-tokenize-a-string-in-c#55680
+  const std::string request(request_buffer);
+  boost::char_separator<char> separator(" ");
+  boost::tokenizer<boost::char_separator<char>> tokens(request, separator);
+  std::cout << "DEBUG: Request tokens in StaticRequestHandler::handleRequest()" << std::endl;
+  for (const auto& t : tokens) {
+    std::cout << t << "." << std::endl;
+  }
+
+  int i = 0;
+  auto curToken = tokens.begin();
+  std::string resourcePath;
+  for (;;) {
+    if (curToken == tokens.end()) {
+      // TODO: tokens is too short
+      std::cout << "DEBUG: tokens in RequestRouter is too short" << std::endl;
+      break;
+    }
+    // We assume the 2nd token is the resource path: GET /path/to/resource.txt
+    if (i == 1) {
+      resourcePath = *curToken;
+    }
+
+    i++;
+    curToken++;
+  }
+
+  std::string request_path = resourcePath;
+  // Cite: Boost Library request_handler.cpp code:
+  // http://www.boost.org/doc/libs/1_49_0/doc/html/boost_asio/
+  // example/http/server/request_handler.cpp
+  if (request_path.empty() || request_path[0] != '/'
+      || request_path.find("..") != std::string::npos)
+  {
+    std::cout << "DEBUG: Bad Request\n" << std::endl;
+    return;
+  }
+  // Cite: Boost Library request_handler.cpp code:
+  // http://www.boost.org/doc/libs/1_49_0/doc/html/boost_asio/
+  // example/http/server/request_handler.cpp
+  std::size_t last_slash_pos = request_path.find_last_of("/");
+  std::string filename = request_path.substr(last_slash_pos + 1);
+  std::size_t last_dot_pos = request_path.find_last_of(".");
+  std::string extension;
+
+  std::string resourceRoot;
+  std::vector<std::string> query = {"server", "location " + request_path.substr(0, last_slash_pos), "root"};
+  server_config.propertyLookUp(query, resourceRoot);
+  // Check if position of last '.' character != end of string AND
+  // position of last '.' comes after position of last '/', then
+  // update extention to contain the file extension
+  // Ex. s = "/bird.png", s[5] = '.', s[0] = '/', extension = "png"
+  if (last_dot_pos != std::string::npos && last_dot_pos > last_slash_pos)
+  {
+    extension = request_path.substr(last_dot_pos + 1);
+  }
+
+  std::string content_type = "Content-Type: " + mime_types::extension_to_type(extension);
+  const std::string response_header = "HTTP/1.1 200 OK\r\n" + content_type + "\r\n\r\n";
+  const size_t header_size = response_header.size();
+  boost::filesystem::path root_path(boost::filesystem::current_path());
+  std::string full_path = root_path.string() + resourceRoot + "/" + filename;
+
+  std::string reply = response_header;
+  std::cout << full_path << std::endl;
+  if (!boost::filesystem::exists(full_path)) {
+    std::cout << "Dispatching 404 handler\n";
```

Comment:
For this log we can probably make it a little more descriptive, like "Dispatching 404 hander: file not found/doesn't exist" something like that